pageserver/src/layered_repository.rs

```diff
@@ -1891,10 +1896,18 @@ pub fn evict_layer_if_needed(conf: &PageServerConf) -> Result<()> {
 
         let timeline = tenant_mgr::get_timeline_for_tenant(tenantid, timelineid)?;
 
-        timeline
+        let mut fsync_paths = timeline
             .upgrade_to_layered_timeline()
             .evict_layer(layer_id)?;
 
+        // FIXME It may be possible for the checkpointer thread to run after `evict_layer` has been
+        // called, but before fsync has finished. This may result in a future disk_consistent_lsn
+        // being written to disk. One way to fix this could be with a checkpointer lock.
```

Comment:
Thoughts on this @hlinnaka

Reply:
hmm, maybe add a new lock on the layer itself, and acquire that in the checkpointer and here.

Or perhaps the list of pending fsyncs should be stored in the LayeredTimeline struct itself, and shared between the checkpointer and this thread. Whenever either the checkpointer or the eviction thread writes out a layer, it would add it to the pending-fsyncs list. At the end of a checkpoint, the checkpointer would wait for the fsyncs to finish, and that would then include any layers evicted by the eviction thread, too. In fact, because we cannot advance disk_consistent_lsn here anyway, there's no to wait for the fsync here, as long as the checkpointer will do it.

(PostgreSQL has a system-wide pending-fsyncs list like that, and it the checkpointer process waits for them all to finish. We're once again copying what is familiar. That's OK, I just find it amusing :-) ).

Reply:
I think we might be able to hold off on this, and not do it if we go the #864 route

Reply:
PR #864 was pushed, so I think we can proceed with this PR as it is.